### PR TITLE
iOS: Make current state compile and run

### LINKED ISSUE #203: State of iOS
Hi,

I'm an iOS engineer who'd love to get my pebble up and running again.

I made the app run on an iOS device and looked into pairing when I noticed that most (if not all) iOS native bridges are not implemented at the moment - is that correct or am I missing something?

### PR DESCRIPTION
**Pull request by q231950**
This pull request makes the app compile and run after checking out the `ios` branch:

- a reference to a non existing file had to be removed
- the _[PermissionControlFlutterBridge](https://github.com/pebble-dev/mobile-app/commit/699e350d26d79f4ca5b0133aabc1dba695a480ed)_ did not fully conform to _PermissionControl_
- fixes an issue where the Flutter engine is run without knowing the background callback handle

Closes #203

### CODE REVIEW

**Review by crc-32 — CHANGES_REQUESTED** [review]
Edit: in reference to https://github.com/pebble-dev/mobile-app/pull/204/commits/bb0d6d4b8cfa1a2a838440623a40562e9472052f
This makes sense to avoid the problem, but it doesn't solve it. The background isolate initializing before there's a handle created should never happen and ignoring this will mean the first launch of the app doesnt run some of the background code, thanks for catching this though!

What should happen is the main function in flutter sets up a background handle and hands it off to native code via [this function](https://github.com/pebble-dev/mobile-app/blob/ios/ios/Runner/bridges/ui/BackgroundSetupFlutterBridge.swift#L10)
Which should be where the first initEngine call is done provided the engine isnt already running (like android does [in that same function](https://github.com/pebble-dev/mobile-app/blob/master/android/app/src/main/kotlin/io/rebble/cobble/bridges/ui/BackgroundSetupFlutterBridge.kt#L25), iOS has an exact analogue for getBackgroundFlutterEngine() in this file that can be used)

So the issue is actually that for some reason I init'd the engine [with the background controller](https://github.com/pebble-dev/mobile-app/blob/ios/ios/Runner/bridges/background/FlutterBackgroundController.swift#L21), which on first launch is before flutter is ready for us when it should just be when flutter sends us the handle like in android.

Feel free to fix that if you get what I'm saying or I will when I have time, not sure how much you know of flutter.

Also thanks a ton for your time, iOS needs more people the most and yet is arguably more important!

**Comment by q231950** [comment]
@crc-32 Thanks for the details, I think I will have a look. 

I did play around with Flutter but have never seen anything at this scale to be honest. Building the Kotlin native XCFramework was another challenge, luckily I understood how these things are supposed to work and what they do but this is definitely a place to see something new 🙂

I'll take this bug and let's see how it goes. Your description of the issue gave me plenty of insight, thanks for taking the time and making it that detailed.

**Comment by q231950** [comment]
@crc-32 I have committed a working draft implementation as a basis for further discussion in 9e33276cd960ce35bd2cf1fff5c2539e377f713f. 

I do not like this solution, it is rather for me to understand how the Flutter Framework and the Pigeons work. I will sketch the solution that I would have liked better but which did not work as expected in the coming days. Maybe you already have some input for this draft implementation. Its main difference to the current implementation is [not to directly start the engine].(https://github.com/pebble-dev/mobile-app/pull/204/files#diff-058d02c5dac990561a86e89151dea8e420febf1fe1e95243e3b7fefdbfa3adf1L23-R27)

Also, I wanted to check if this is the furthest a pairing on iOS can go at the moment or if I am holding it wrong:

**Comment by q231950** [comment]
Hi @crc-32 ,

with 77d2bd6d33c454138add48517f01b36df884498f I was able to wait for the background handle without using a notification. This makes me happier. I am not super familiar to PromiseKit but it seemed the proper way to handle this 'waiting', looking at the rest of the code. I had tried with closures without luck before but closures don't seem appropriate here anyways.

Please have a look at the changes once you find the time and let me know what you think. If this is a completely different approach of what you had in mind don't hesitate and tell me as well - I can take it 😉

Looking at my stashes which accumulated the last couple of days I believe that [this](https://github.com/pebble-dev/mobile-app/blob/ios/ios/Runner/bridges/ui/ConnectionControlBridge.swift#L11-L14) doesn't work because it will only call once with some state like 'waiting to enable bluetooth' and all successive state changes will be missed. It's possible that I am not getting the full picture though.

**Comment by crc-32** [comment]
> with [77d2bd6](https://github.com/pebble-dev/mobile-app/commit/77d2bd6d33c454138add48517f01b36df884498f) I was able to wait for the background handle without using a notification. This makes me happier. I am not super familiar to PromiseKit but it seemed the proper way to handle this 'waiting', looking at the rest of the code. I had tried with closures without luck before but closures don't seem appropriate here anyways.
> 
> Please have a look at the changes once you find the time and let me know what you think. If this is a completely different approach of what you had in mind don't hesitate and tell me as well - I can take it 😉

This looks good to me! as i commented on the commit you can get rid of the unused return with `cauterize()`

> Looking at my stashes which accumulated the last couple of days I believe that [this](https://github.com/pebble-dev/mobile-app/blob/ios/ios/Runner/bridges/ui/ConnectionControlBridge.swift#L11-L14) doesn't work because it will only call once with some state like 'waiting to enable bluetooth' and all successive state changes will be missed. It's possible that I am not getting the full picture though.

That's almost certainly the issue yes

File: ios/Runner/bridges/background/FlutterBackgroundController.swift [hunk 1]
```diff
@@ -35,33 +29,35 @@ class FlutterBackgroundController: NSObject, BackgroundControl {
...
-            }else {
-                initEngine().done { initedEngine in
-                    self.engine = initedEngine
-                    seal.fulfill(self.engine)
-                }.catch { error in
-                    DDLogError("Error initializing background engine: \(error.localizedDescription)")
-                    seal.reject(error)
-                }
+    func getBackgroundFlutterEngine() -> Promise<FlutterEngine> {
+        Promise { backgroundEngineGetters.append($0) }
```
Comment (q231950) [inline ios/Runner/bridges/background/FlutterBackgroundController.swift:33]: This still needs to fulfill the promise immediately when the engine is set up.

File: ios/Runner/bridges/FlutterBridgeSetup.swift [hunk 2]
```diff
@@ -18,11 +20,16 @@ class FlutterBridgeSetup {
...
+
+        backgroundSetupBridge.waitForBackgroundHandle().done {
+            FlutterBackgroundController.shared.setupEngine($0)
+        }.cauterize()
+
+        BackgroundSetupControlSetup(binaryMessenger, backgroundSetupBridge)
     }
     
     static func createUIBridges(binaryMessenger: FlutterBinaryMessenger) {
-        BackgroundSetupControlSetup(binaryMessenger, BackgroundSetupFlutterBridge())
```
Comment (crc-32) [inline ios/Runner/bridges/FlutterBridgeSetup.swift:25]: `createBackgroundBridges` is for all the (unimplemented currently) bridges which run on the background flutter isolate, since background setup runs before this its considered a UI bridge even if its kind of a misnomer so it still should be put in here (or in its own method maybe) as otherwise we end up initialising it again for the background code when its not needed.
tldr please move it back into `createUIBridges` (i get the confusion though!) and don't use `createBackgroundBridges` with the main flutter engine [here](https://github.com/pebble-dev/mobile-app/pull/204/files#diff-2d8ca8a3da302654c934d7f3be4c2f6c8edc7788f2b8ed0178730d6d703ebe72R29)

**Review by crc-32 — CHANGES_REQUESTED** [review]
Almost all of this is perfect, thank you!
Just one quick thing I missed from earlier.

File: ios/Runner/bridges/FlutterBridgeSetup.swift (hunk 2 above)
Comment (q231950) [reply ios/Runner/bridges/FlutterBridgeSetup.swift:25]: Alrighty, I (kind of) understood what you are saying and pushed a suggestion, having `backgroundSetupBridge` as a private var. 
It is confusing with the names, I tried to give it some more structure with the 2 code marks. Tell me what you think - I am looking forward to get this in. To you it may have been a small thing but for me it was a great entry point to get into the project. Thanks for all the explanations 🙂

**Review by q231950 — COMMENTED** [review]
@crc-32  please have another look once you have a moment. Please also have a look at my last commit where I used the `typealias CallbackHandle = Int64` more consistently. I know that this kind of style can be controversial, just let me know if I should keep it or leave it.

**Review by crc-32 — APPROVED** [review]

**Comment by crc-32** [comment]
That all looks fine, thank you!
